query-engine/core/src/telemetry/helpers.rs

```diff
@@ -0,0 +1,112 @@
+    }
+}
+
+// set the parent context and return the traceparent
+pub fn set_parent_context_from_json_str(span: &Span, trace: &str) -> Option<String> {
+    let trace: HashMap<String, String> = serde_json::from_str(trace).unwrap_or_default();
+    let trace_id = trace.get("traceparent").map(String::from);
+    let cx = opentelemetry::global::get_text_map_propagator(|propagator| propagator.extract(&trace));
+    span.set_parent(cx);
+    trace_id
+}
+
+pub fn set_span_link_from_traceparent(span: &Span, traceparent: Option<String>) {
+    if let Some(traceparent) = traceparent {
+        let trace: HashMap<String, String> = HashMap::from([("traceparent".to_string(), traceparent)]);
+        let cx = opentelemetry::global::get_text_map_propagator(|propagator| propagator.extract(&trace));
+        let context_span = cx.span();
+        span.add_link(context_span.span_context().clone());
+    }
+}
+
+pub fn get_trace_id_from_context(context: &Context) -> TraceId {
+    let context_span = context.span();
+    context_span.span_context().trace_id()
+}
+
+pub enum QueryEngineLogLevel {
+    FromEnv,
+    Override(String),
+}
+
+impl QueryEngineLogLevel {
+    fn level(self) -> Option<String> {
+        match self {
+            Self::FromEnv => std::env::var("QE_LOG_LEVEL").ok(),
+            Self::Override(l) => Some(l),
+        }
+    }
+}
+
+#[rustfmt::skip]
+pub fn env_filter(log_queries: bool, qe_log_level: QueryEngineLogLevel) -> EnvFilter {
+    let mut filter = EnvFilter::from_default_env()
+        .add_directive("tide=error".parse().unwrap())
+        .add_directive("tonic=error".parse().unwrap())
+        .add_directive("h2=error".parse().unwrap())
+        .add_directive("hyper=error".parse().unwrap())
+        .add_directive("tower=error".parse().unwrap());
+
+    if let Some(level) = qe_log_level.level() {
+        filter = filter
+            .add_directive(format!("query_engine={}", &level).parse().unwrap())
+            .add_directive(format!("query_core={}", &level).parse().unwrap())
+            .add_directive(format!("query_connector={}", &level).parse().unwrap())
+            .add_directive(format!("sql_query_connector={}", &level).parse().unwrap())
+            .add_directive(format!("mongodb_query_connector={}", &level).parse().unwrap());
+    }
+
+    if log_queries {
+        filter = filter
+            .add_directive("quaint[{is_query}]=trace".parse().unwrap())
+            .add_directive("mongodb_query_connector=debug".parse().unwrap());
+    }
+
+    filter
+}
+
+pub fn user_facing_span_only_filter(meta: &Metadata) -> bool {
+    if !meta.is_span() {
+        return false;
+    }
+
+    user_facing_filter(meta)
+}
+
+pub fn user_facing_filter(meta: &Metadata) -> bool {
+    if *SHOW_ALL_TRACES {
+        return true;
+    }
+
+    if meta.fields().iter().any(|f| f.name() == "user_facing") {
+        return true;
+    }
+
+    meta.target() == "quaint::connector::metrics" && meta.name() == "quaint:query"
+}
```

Comment:
I see a lot of `pub fn` items in this file and elsewhere, and that is wholesale reexported as public API to consumers of the crate. Does this all need to be public? It make the API of a crate smaller and easier to understand if internal items are `pub(crate)` instead.

Reply:
Right, I didn't restrict visibility of any of those, and just moved it. I will check and restrict when possible.